torch/_dynamo/symbolic_convert.py

```diff
@@ -1627,6 +1627,41 @@ def FOR_ITER(self, inst):
             self.jump(inst)
 
     def _raise_exception_variable(self, inst):
+        def set_context_recursive(val, prev_idx):
+            if (ctx := val.__context__) and type(ctx) is not ConstantVariable:
+                return val
+            if len(self.exn_vt_stack) + prev_idx > 0:
+                prev = self.exn_vt_stack[prev_idx]
+                set_context_recursive(prev, prev_idx - 1)
+                val.set_context(prev)
+            return val
+
+        def break_context_reference_cycle(val):
+            # See test_exceptions::test_raise_does_not_create_context_chain_cycle
+            # Based on https://github.com/python/cpython/blob/e635bf2e49797ecb976ce45a67fce2201a25ca68/Python/errors.c#L207-L228
+            # As noted on CPython, this is O(chain length) but the context chains
+            # are usually very small
```

Comment:
Er what is going on here? Why is this needed?

Reply:
When setting the __context__ attribute of an exception, there's a chance this creates a reference cycle. This piece of code is to detect and remove such cases.

Reply:
how exactly are you removing a reference cycle?

Reply:
We don't set the context if it creates a cycle. There are a couple of tests that checks for this behavior. The algorithm I implemented is a straight translation of what CPython do in this case.

Reply:
interesting

Reply:
I didn't quite answer how we detect cycles. But it works as follows:
(i) Set the `__context__` attribute
(ii) Detect if there's a cycle by using floyd's algorithm ([ref](https://www.geeksforgeeks.org/floyds-cycle-finding-algorithm/))
(iii) If yes, remove the context.